backends/candle/src/models/qwen3.rs

```diff
@@ -18,3 +22,577 @@ pub struct Qwen3Config {
+        let gate_up_proj = Linear::new(gate_up_proj_weight, None, None);
+
+        let down_proj_weight = vb
+            .pp("down_proj")
+            .get((config.hidden_size, intermediate_size), "weight")?;
+        let down_proj = Linear::new(down_proj_weight, None, None);
+
+        Ok(Self {
+            gate_up_proj,
+            down_proj,
+            intermediate_size,
+            span: tracing::span!(tracing::Level::TRACE, "mlp"),
+        })
+    }
+
+    pub fn forward(&self, hidden_states: &Tensor) -> Result<Tensor> {
+        let _enter = self.span.enter();
+
+        let gate_up_states = self.gate_up_proj.forward(hidden_states)?;
+        let gate_states = gate_up_states.narrow(D::Minus1, 0, self.intermediate_size)?;
+        let up_states = gate_up_states.narrow(D::Minus1, self.intermediate_size, self.intermediate_size)?;
+
+        let gate_states = gate_states.silu()?;
+        self.down_proj.forward(&(gate_states * up_states)?)
+    }
+}
+
+struct Qwen3Layer {
+    attention: Qwen3Attention,
+    mlp: Qwen3MLP,
+    input_layer_norm: RMSNorm,
+    post_attention_layer_norm: RMSNorm,
+
+    span: tracing::Span,
+}
+
+impl Qwen3Layer {
+    pub fn load(vb: VarBuilder, config: &Qwen3Config) -> Result<Self> {
+        let attention = Qwen3Attention::load(vb.pp("self_attn"), config)?;
+        let mlp = Qwen3MLP::load(vb.pp("mlp"), config)?;
+
+        let input_layer_norm = RMSNorm::load(
+            vb.pp("input_layernorm"),
+            config.hidden_size,
+            config.rms_norm_eps,
+        )?;
+        let post_attention_layer_norm = RMSNorm::load(
+            vb.pp("post_attention_layernorm"),
+            config.hidden_size,
+            config.rms_norm_eps,
+        )?;
+
+        Ok(Self {
+            attention,
+            mlp,
+            input_layer_norm,
+            post_attention_layer_norm,
+            span: tracing::span!(tracing::Level::TRACE, "layer"),
+        })
+    }
+
+    pub fn forward(
+        &self,
+        hidden_states: &Tensor,
+        attention_bias: Option<&Tensor>,
+        cos: &Tensor,
+        sin: &Tensor,
+    ) -> Result<Tensor> {
+        let _enter = self.span.enter();
+
+        let (normed_hidden_states, res) = self.input_layer_norm.forward(hidden_states, None)?;
+        let attn_output = self.attention.forward(&normed_hidden_states, attention_bias, cos, sin)?;
+        let (normed_attn_res_output, attn_res) = self.post_attention_layer_norm.forward(&attn_output, Some(&res))?;
+        let mlp_output = self.mlp.forward(&normed_attn_res_output)?;
+
+        // Add residual connections
+        let output = (&mlp_output + &attn_res)?;
+        Ok(output)
+    }
+}
+
+pub struct Qwen3Model {
+    embeddings: Embedding,
+    layers: Vec<Qwen3Layer>,
+    norm: RMSNorm,
+    rotary_cache: (Tensor, Tensor),
+    rotary_dim: usize,
+    pool: Pool,
+    pub device: Device,
+    num_attention_heads: usize,
```

Comment:
Is this really needed? Isn't it more reliable to just `layers[0].attention.attention_head_size` within the `load` method?